### ISSUE TITLE
add "pct_change" feature for Series// groupby ops

### ISSUE THREAD

**Issue by yutiansut** [pola-rs/polars#2184]
something like this 

```rust
    trait pct{
       fn pct_change(&self, n: usize,) -> Series ;
    }
    impl pct for Series{
        fn pct_change(&self, n: usize,) -> Series {
            self / &self.shift(n as i64)
        }
    }
```

**Comment by yutiansut** [comment]
this's something very usefull in financial calc

python mode like 

```
df.groupby( 'order_book_id').close.pct_change(N)
```

**Comment by yutiansut** [comment]
@dbr @ritchie46

**Comment by yutiansut** [comment]
something about groupby method: 

the first i thought is something like



```rust
df.groupby("code").unwarp().select("close").pct_change(N)
```

however its wrong since the select-> aggmethod is something agg chrunk array => to a single value

like sum ops, sum[chrunk array ] -> sumvalue - >  [value] -> make it as a series


however the pct_change or diff or shift methods may not agg the data

so there may be another way to transform the data inside a groups

maybe is   something like ```data.groupby('code').unwrap().apply(| x| Ok(dataframe)).unwrap()```

however there is still error 

1. the groupby-> apply methods ask for a Result of dataframe not the mut dataframe , so the data inside can not be changed itself

2. since we do not use groupby , the mut dataframe's apply methods seems still can't change the columns

```rust 
fn diff_for_series(data: &Series)-> Series{
    data.shift(1)
}
df.apply("close", diff_for_series);
```

this will raise a exception about ```all columns in a record batch must have same length``` error

**Comment by yutiansut** [comment]
the idea methods is something like Series with index, and adds methods about that

the index in dataframe[pandas] is somewhere usefull cause it take extra message in index while we just handle a single columns[without handle the whole dataframe]

since the polars doesn't support the index methods, there maybe to seek the answer from the dataframe( Series in polars is just something like a 1D-array just with names)  

however the dataframe in polars can't support for groupby and change the columns inside ?  maybe misunderstood  

is there any solutions about that?

**Comment by ritchie46** [comment]
With polars lazy or in python we can compute within groups

```python
(df.groupby("foo")
   .agg([
    pl.col("time") / pl.col("time").shift(10).alias("pct_change")
])
```

We can make a `pct_change` expression, but it is really easy to compose from existing expressions already.

**Comment by yutiansut** [comment]
i just success with
```
 trait pct {
        fn pctchange(&self, n: usize) -> Series;
    }
    impl pct for Series {
        fn pctchange(&self, n: usize) -> Series {
            self / &self.shift(n as i64)
        }
    }

    fn closepctchange(close: &Series) -> Series {
        close.pctchange(1)
    }
    let rank2 = rank
        .groupby("order_book_id")
        .unwrap()
        .apply(|mut x| {
            let res = x
                .sort("date", false)
                .unwrap()
                .apply("close", closepctchange)
                .unwrap()
                .clone();
            //println!("rank {}", rank["close"]);
            Ok(res)
        })
        .unwrap()
        .sort("date", false)
        .unwrap()
        .drop_nulls(Some(&["close".to_string()]))
        .unwrap();
```

**Comment by yutiansut** [comment]
@ritchie46 am i something wrong with this 

```rust
    let rank3 = rank
        .lazy()
        .groupby([col("order_book_id")])
        .agg([col("close") / col("close").shift(1).alias("pctchange")])
        .collect()
        .unwrap();

```
this is the error

ComputeError("this binary expression is not an aggregation: [(close) Divide (SHIFT close by 1 AS pctchange)]\n                pherhaps you should add an aggregation like, '.sum()', '.min()', '.mean()', etc.

**Comment by ritchie46** [comment]
> @ritchie46 am i something wrong with this
> 
> ```rust
>     let rank3 = rank
>         .lazy()
>         .groupby([col("order_book_id")])
>         .agg([col("close") / col("close").shift(1).alias("pctchange")])
>         .collect()
>         .unwrap();
> ```
> 
> this is the error
> 
> ComputeError("this binary expression is not an aggregation: [(close) Divide (SHIFT close by 1 AS pctchange)]\n pherhaps you should add an aggregation like, '.sum()', '.min()', '.mean()', etc.

Finish with a list aggregation 

`col("close").shift(1).list().alias("pctchange")])`

**Comment by yutiansut** [comment]
it work with 
```rust
    let rank3 = rank
        .lazy()
        .groupby([col("order_book_id")])
        .agg([(col("close") / col("close").shift(1)).list().alias("pctchange")])
        .collect()
        .unwrap();
```

and the result 
```
rank lazy shape: (200, 2)
┌───────────────┬─────────────────────────────────────┐
│ order_book_id ┆ pctchange                           │
│ ---           ┆ ---                                 │
│ str           ┆ list [f32]                          │
╞═══════════════╪═════════════════════════════════════╡
│ 300770.XSHE   ┆ [1.1483378410339355, 0.777705907... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 300223.XSHE   ┆ [4.656919956207275, 6.3425574302... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 002685.XSHE   ┆ [0.07251287996768951, 0.05450732... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 002312.XSHE   ┆ [0.16866889595985413, 0.18521486... │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...           ┆ ...                                 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤

```

however it can't make sure the sort of date and it returns something like a sql query

this is the result from https://github.com/pola-rs/polars/issues/2184#issuecomment-1001709147
```
┌────────────┬───────────────┬────────────────────┐
│ date       ┆ order_book_id ┆ close              │
│ ---        ┆ ---           ┆ ---                │
│ str        ┆ str           ┆ f32                │
╞════════════╪═══════════════╪════════════════════╡
│ 2019-01-03 ┆ 300618.XSHE   ┆ 1.0031566619873047 │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2019-01-03 ┆ 300038.XSHE   ┆ 0.9717868566513062 │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2019-01-03 ┆ 300628.XSHE   ┆ 0.9834636449813843 │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2019-01-03 ┆ 300613.XSHE   ┆ 0.996540904045105  │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...        ┆ ...           ┆ ...                │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2021-12-08 ┆ 300915.XSHE   ┆ 0.9981292486190796 │

```

**Comment by yutiansut** [comment]
Lazy model is something like datafusion? power by a logicalplan and physicalplan to query and transform the data. 

So which is better for a pure rust like function or sql execute [lazy] function?

**Comment by ritchie46** [comment]
closed by #2194 

> Lazy model is something like datafusion? power by a logicalplan and physicalplan to query and transform the data.

Yes, the user guide explains more about lazy: https://pola-rs.github.io/polars-book/user-guide/index.html
